http://git.savannah.gnu.org/cgit/grep.git/commit/?id=7f96aa773b5967bbb3c8476f8614fd04b3fc1362

From 7f96aa773b5967bbb3c8476f8614fd04b3fc1362 Mon Sep 17 00:00:00 2001
From: Jim Meyering <meyering@fb.com>
Date: Fri, 01 Nov 2013 03:20:30 +0000
Subject: grep: fix regression involving \s and \S

Commit v2.14-40-g01ec90b made \s and \S work with multi-byte
characters, but it made it so any use like \s*, \s+, \s?, \s{3}
would malfunction in a multi-byte locale.
* src/dfa.c (lex): Also reset laststart.
* tests/backslash-s-and-repetition-operators: New file.
* tests/Makefile.am (TESTS): Add it.
* NEWS (Bug fixes): Mention it.
* THANKS: Update.
Reported by Mirraz Mirraz in http://bugs.gnu.org/15773.
---
--- a/NEWS
+++ b/NEWS
@@ -39,6 +39,11 @@
   it.  It is only available if the corresponding support in the PCRE
   library is detected when grep is compiled.
 
+  The fix to make \s and \S work with multi-byte white space broke
+  the use of each shortcut whenever followed by a repetition operator.
+  For example, \s*, \s+, \s? and \s{3} would all malfunction in a
+  multi-byte locale.  [bug introduced in grep-2.15]
+
 
 * Noteworthy changes in release 2.14 (2012-08-20) [stable]
 
--- a/THANKS
+++ b/THANKS
@@ -66,6 +66,7 @@
 Michael Aichlmayr          <mikla@nx.com>
 Mike Frysinger             <vapier@gentoo.org>
 Miles Bader                <miles@ccs.mt.nec.co.jp>
+Mirraz Mirraz              <mirraz1@rambler.ru>
 Nelson H. F. Beebe         <beebe@math.utah.edu>
 Olaf Kirch                 <okir@ns.lst.de>
 Paolo Bonzini              <bonzini@gnu.org>
--- a/src/dfa.c
+++ b/src/dfa.c
@@ -1473,6 +1473,7 @@
 
           POP_LEX_STATE ();
 
+          laststart = 0;
           return lasttok;
 
         case 'w':
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -36,6 +36,7 @@
   backref					\
   backref-multibyte-slow			\
   backref-word					\
+  backslash-s-and-repetition-operators		\
   backslash-s-vs-invalid-multitype		\
   big-hole					\
   big-match					\
--- /dev/null
+++ b/tests/backslash-s-and-repetition-operators
@@ -0,0 +1,36 @@
+#! /bin/sh
+# Ensure that \s and \S work with repetition operators.
+#
+# Copyright (C) 2013 Free Software Foundation, Inc.
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.
+
+. "${srcdir=.}/init.sh"; path_prepend_ ../src
+
+require_en_utf8_locale_
+
+fail=0
+
+for loc in en_US.UTF-8 C; do
+  echo locale=$loc
+  LC_ALL=$loc
+  export LC_ALL
+
+  printf ' \n' > in || framework_failure_
+
+  for re in '\s\+' '\s*' '\s\?' '\s\{1\}'; do
+    grep "^$re\$" in > out || fail=1
+    compare in out || fail=1
+  done
+
+  printf 'X\n' > in || framework_failure_
+
+  for re in '\S\+' '\S*' '\S\?' '\S\{1\}'; do
+    grep "^$re\$" in > out || fail=1
+    compare in out || fail=1
+  done
+done
+
+Exit $fail
